### PR TITLE
ARROW-12110: [Java] Implement ZSTD compression

### PR DESCRIPTION
**Pull request by emkornfield**
* Add ZSTD codec implementation.
* Removes dependency on netty by using ArrowBuf methods.
* Updates docs and archery test (will wait on CI to run archery if that fails will do more debugging.)

### CODE REVIEW

File: dev/archery/archery/integration/runner.py [hunk 1]
```diff
@@ -137,8 +137,6 @@ def _gold_tests(self, gold_dir):
                 skip.add("Go")
                 skip.add("JS")
                 skip.add("Rust")
-            if name == 'zstd':
```
Comment (nevi-me) [inline dev/archery/archery/integration/runner.py:140]: A downside to removing this comment completely is that someone might not easily figure out how to skip tests based on name (when working on another implementation).

Do you mind adding a comment to the effect that if an implementer is skipping one of the compression formats, they could use `if name == 'format'`?

Comment (emkornfield) [reply dev/archery/archery/integration/runner.py:140]: Added a link back to this PR>

**Comment by emkornfield** [comment]
I ran archery tests for java locally and zstd seems to pass with this PR.  I'm not sure how these changes could have affected the archery/crossbow tests.  

@dianaclarke I think you might have been the last person to touch this area?  Any thoughts?

**Comment by dianaclarke** [comment]
> I ran archery tests for java locally and zstd seems to pass with this PR. I'm not sure how these changes could have affected the archery/crossbow tests.
> 
> @dianaclarke I think you might have been the last person to touch this area? Any thoughts?

I'm pretty new to Arrow, and I'm not quite sure I follow. What's failing?

The `Archery & Crossbow / Archery Unittests and Crossbow Check Config` check passed.

Are you referring to the `Integration / AMD64 Conda Integration Test (pull_request)` check? It appears to be out of disk space.

```
npm ERR! nospc ENOSPC: no space left on device, write
npm ERR! nospc There appears to be insufficient space on your system to finish.
```

I'm probably too n00b to be of much help on this issue, but good luck!

That said, I do hope to find some time to collaborate on the following PR with you all.

    ARROW-10031: [CI][Java] Support Java benchmark in Archery
    https://github.com/apache/arrow/pull/8210

I would be fun to get these Java benchmarks hooked up to Conbench. I have a blog post coming soon that better explains the Arrow/Archery/Conbench integration. Until then:

    https://conbench.ursa.dev/
    https://github.com/ursacomputing/conbench
    https://github.com/ursacomputing/benchmarks

Cheers & stay safe!

**Comment by emkornfield** [comment]
@dianaclarke hmm, it looks like "Archery & Crossbow / Archery Unittests and Crossbow Check Config (pull_request" passed this time, it seems like it might be flaky (i don't know that there is a great way to see old runs).  Yeah, I think the integration test failure is completely unrelated and there is already an open JIRA for it.

File: java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import com.github.luben.zstd.Zstd;
+
+/**
+ * Compression codec for the ZSTD algorithm.
+ */
+public class ZstdCompressionCodec extends AbstractCompressionCodec {
+
+  @Override
+  protected ArrowBuf doCompress(BufferAllocator allocator, ArrowBuf uncompressedBuffer) {
+    long maxSize = Zstd.compressBound(uncompressedBuffer.writerIndex());
```
Comment (liyafan82) [inline java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java:35]: Is it possible to cause too much waste of memroy in the worst case?

File: java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    long dstSize = CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH + maxSize;
+    ArrowBuf compressedBuffer = allocator.buffer(dstSize);
+    long bytesWritten = Zstd.compressUnsafe(
+                          compressedBuffer.memoryAddress() + CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH, dstSize,
+                          /*src*/uncompressedBuffer.memoryAddress(), /*srcSize=*/uncompressedBuffer.writerIndex(),
+                          /*level=*/3);
```
Comment (liyafan82) [inline java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java:41]: Should we make the level configurable?

File: java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+  @Override
+  protected ArrowBuf doDecompress(BufferAllocator allocator, ArrowBuf compressedBuffer) {
+    long decompressedLength = readUncompressedLength(compressedBuffer);
+    ArrowBuf uncompressedBuffer = allocator.buffer(decompressedLength);
+    long decompressedSize = Zstd.decompressUnsafe(uncompressedBuffer.memoryAddress(), decompressedLength,
+          /*src=*/compressedBuffer.memoryAddress() + CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH,
+          compressedBuffer.writerIndex() - CompressionUtil.SIZE_OF_UNCOMPRESSED_LENGTH);
+    if (Zstd.isError(decompressedSize)) {
+      uncompressedBuffer.close();
```
Comment (liyafan82) [inline java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java:58]: We can place the `close` statement in a finally block?

**Comment by liyafan82** [comment]
I am curious to know about the performance of this codec (I suppose it should be good, as there are no on-heap/off-heap data copies). 
Maybe we can run some benchmarks after ARROW-11901 checks in the benchmarks.

File: java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java (hunk 3 above)
Comment (emkornfield) [reply java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java:41]: maybe at some point, the default is 3.  LZ4 also has a compression level which I don't think we are setting.

File: java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java (hunk 2 above)
Comment (emkornfield) [reply java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java:35]: I think this is just a fact of life  with compression algorithms.  There are potentially ways of mitigating memory waste if changed the APIs around a little bit (trying to pack more data into a single ArrowBuf) or consolidating buffers into a larger buffer if compression level is high.  But this is the guidance the documented methods give for sizing buffers.

**Comment by emkornfield** [comment]
Yes, I expect it to be good.  I'm going to try to run some end-to-end experiments and I can report back numbers.

File: java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java (hunk 3 above)
Comment (emkornfield) [reply java/compression/src/main/java/org/apache/arrow/compression/ZstdCompressionCodec.java:41]: I'll open up a JIRA to track configuration.

**Comment by emkornfield** [comment]
@dianaclarke https://github.com/apache/arrow/pull/9822/checks?check_run_id=2219940455

**Comment by dianaclarke** [comment]
> @dianaclarke https://github.com/apache/arrow/pull/9822/checks?check_run_id=2219940455

Ah, now I see. I didn't write that test. I think @bkietz did, but I can try to find some time this week to dig into it.

**Comment by dianaclarke** [comment]
> > @dianaclarke https://github.com/apache/arrow/pull/9822/checks?check_run_id=2219940455
> 
> Ah, now I see. I didn't write that test. I think @bkietz did, but I can try to find some time this week to dig into it.

I see what the issue is. That test is assuming order, when it can't. I'll create a ticket and submit a PR shortly.

[Developer][Archery] Flaky test: test_static_runner_from_json
https://issues.apache.org/jira/browse/ARROW-12145

**Comment by emkornfield** [comment]
@liyafan82 I opened https://issues.apache.org/jira/browse/ARROW-12163 to track making compression levels configurable, any other concerns?

**Comment by liyafan82** [comment]
> @liyafan82 I opened https://issues.apache.org/jira/browse/ARROW-12163 to track making compression levels configurable, any other concerns?

@emkornfield My only concern is performance. If we are going to address it in ARROW-11901, this PR is ready for merge IMO.

**Comment by emkornfield** [comment]
> @emkornfield My only concern is performance. If we are going to address it in ARROW-11901, this PR is ready for merge IMO.

Yes, I think functionality should come first and we can measure and iterate on performance.  Thank you for your review.

**Comment by emkornfield** [comment]
going to merge.

**Pull request closed without merging**